### PR TITLE
 [core] fix exit handling of FiberState threads

### LINKED ISSUE #45656: [Core] async actors do not terminate cleanly with `__ray_terminate__`
### What happened + What you expected to happen

I expected atexit handlers from the actors to be called when terminating an actor with `__ray_terminate__ `.

The error seems to happen only when the below is true - 
* actor is async
* a remote method on the actor has been called at least once
* the actor is terminated from a separate shell/proess than the one which created it.

Exception in the actor -

```
2024-05-31 13:51:57,629 E 16341 16341] logging.cc:101: Unhandled exception: St12system_error. what(): Invalid argument
[2024-05-31 13:51:57,652 E 16341 16341] logging.cc:108: Stack trace: 
 /home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x101a47a) [0x7f642089447a] ray::operator<<()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x101cf38) [0x7f6420896f38] ray::TerminateHandler()
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xae20c) [0x7f641f6fc20c]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xae277) [0x7f641f6fc277]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xae4d8) [0x7f641f6fc4d8]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(_ZSt20__throw_system_errori+0x96) [0x7f641f6f383c] std::__throw_system_error()
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xdc30e) [0x7f641f72a30e]
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x7f13d9) [0x7f642006b3d9] ray::core::ConcurrencyGroupManager<>::Stop()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core28CoreWorkerDirectTaskReceiver4StopEv+0x7b) [0x7f6420051e6b] ray::core::CoreWorkerDirectTaskReceiver::Stop()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorker8ShutdownEv+0x230) [0x7f64200247a0] ray::core::CoreWorker::Shutdown()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0xa5acee) [0x7f64202d4cee] EventTracker::RecordExecution()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0xa540de) [0x7f64202ce0de] std::_Function_handler<>::_M_invoke()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0xa54556) [0x7f64202ce556] boost::asio::detail::completion_handler<>::do_complete()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x11041ab) [0x7f642097e1ab] boost::asio::detail::scheduler::do_run_one()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x1105b29) [0x7f642097fb29] boost::asio::detail::scheduler::run()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x1106232) [0x7f6420980232] boost::asio::io_context::run()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorker20RunTaskExecutionLoopEv+0xcd) [0x7f641fff396d] ray::core::CoreWorker::RunTaskExecutionLoop()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core21CoreWorkerProcessImpl26RunWorkerTaskExecutionLoopEv+0x8c) [0x7f6420038c2c] ray::core::CoreWorkerProcessImpl::RunWorkerTaskExecutionLoop()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core17CoreWorkerProcess20RunTaskExecutionLoopEv+0x1d) [0x7f6420038ddd] ray::core::CoreWorkerProcess::RunTaskExecutionLoop()
/home/container_user/.local/lib/python3.10/site-packages/ray/_raylet.so(+0x5c2a07) [0x7f641fe3ca07] __pyx_pw_3ray_7_raylet_10CoreWorker_7run_task_loop()
ray::DummyActor(+0x15e6c7) [0x564466ea56c7]
ray::DummyActor(_PyEval_EvalFrameDefault+0x8ac) [0x564466e8a45c] _PyEval_EvalFrameDefault
ray::DummyActor(_PyFunction_Vectorcall+0x7c) [0x564466ea19fc] _PyFunction_Vectorcall
ray::DummyActor(_PyEval_EvalFrameDefault+0x8ac) [0x564466e8a45c] _PyEval_EvalFrameDefault
ray::DummyActor(+0x13f9c6) [0x564466e869c6]
ray::DummyActor(PyEval_EvalCode+0x86) [0x564466f7c256] PyEval_EvalCode
ray::DummyActor(+0x260108) [0x564466fa7108]
ray::DummyActor(+0x2599cb) [0x564466fa09cb]
ray::DummyActor(+0x25fe55) [0x564466fa6e55]
ray::DummyActor(_PyRun_SimpleFileObject+0x1a8) [0x564466fa6338] _PyRun_SimpleFileObject
ray::DummyActor(_PyRun_AnyFileObject+0x43) [0x564466fa5f83] _PyRun_AnyFileObject
ray::DummyActor(Py_RunMain+0x2be) [0x564466f98a5e] Py_RunMain
ray::DummyActor(Py_BytesMain+0x2d) [0x564466f6f02d] Py_BytesMain
/usr/lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f6421565d90]
/usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f6421565e40] __libc_start_main
ray::DummyActor(_start+0x25) [0x564466f6ef25] _start

*** SIGABRT received at time=1717163517 on cpu 2 ***
PC: @     0x7f64215d29fc  (unknown)  pthread_kill
    @     0x7f642157e520  (unknown)  (unknown)
[2024-05-31 13:51:57,652 E 16341 16341] logging.cc:365: *** SIGABRT received at time=1717163517 on cpu 2 ***
[2024-05-31 13:51:57,652 E 16341 16341] logging.cc:365: PC: @     0x7f64215d29fc  (unknown)  pthread_kill
[2024-05-31 13:51:57,652 E 16341 16341] logging.cc:365:     @     0x7f642157e520  (unknown)  (unknown)
Fatal Python error: Aborted

Stack (most recent call first):
  File "/home/container_user/.local/lib/python3.10/site-packages/ray/_private/worker.py", line 876 in main_loop
  File "/home/container_user/.local/lib/python3.10/site-packages/ray/_private/workers/default_worker.py", line 289 in <module>
```

### Versions / Dependencies

2.8.0 / 2.23.0

### Reproduction script

1. start a local ray cluster `ray start --head`

2. start an actor - 

```python
import atexit
import ray
import asyncio


class DummyActor:

    def __init__(self):

        print("actor created")
        atexit.register(self.print_exit)

    def print_exit(self):

        print("exiting actor")


    async def api(self):

        print("api called")


actor = ray.remote(DummyActor).options(lifetime="detached", namespace="test", name="test").remote()
ray.get(actor.api.remote())
```
3. try to stop the actor - 

```
import ray
actor_handle = ray.get_actor(namespace="test", name="test")
ray.wait([actor_handle.__ray_terminate__.remote()])
```

I don't see the print "exiting actor" in the logs. There are exceptions in the stderr of the actor are shown above.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by hongchaodeng**
## Why are these changes needed?

Currently `fiber_runner_thread_.detach()` is called in `Fiber.Join()`. But `Fiber.Join()` could be called multiple times. For example, `ConcurrencyGroupManager` could call `Join()` the same number of times as the number of workers. This could result in double `detach()` failure.

`detach()` **should** be called after `fiber_runner_thread_` was created. That's the right pattern to use std thread.

## Related issue number

Fix https://github.com/ray-project/ray/issues/45656

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rynewang** [comment]
We can add tests to `src/ray/core_worker/test/fiber_state_test.cc` by calling Stop() and Join() twice and it did not crash

**Comment by hongchaodeng** [comment]
@jjyao @rynewang Fixed!

File: src/ray/core_worker/fiber.h [hunk 1]
```diff
@@ -140,6 +140,8 @@ class FiberState {
         std::this_thread::sleep_for(std::chrono::hours(1));
       }
     });
+
+    fiber_runner_thread_.detach();
```
Comment (rynewang) [inline src/ray/core_worker/fiber.h:144]: can we just remove this member variable and make it a local object just in ctor?

**Review by rynewang — APPROVED** [review]

File: src/ray/core_worker/fiber.h [hunk 2]
```diff
@@ -155,7 +157,6 @@ class FiberState {
 
   void Join() {
     fiber_stopped_event_->Wait();
-    fiber_runner_thread_.detach();
```
Comment (jjyao) [inline src/ray/core_worker/fiber.h:158]: Currently we are relying on the behavior that `channel_.close();` and `fiber_stopped_event_->Wait();` can be called multiple times, which I checked is true. But this seems fragile and rely on the underlying behavior of these libraries. Can we just have our own `stopped_` and `joined_`  flags and early return if they are already called? Thoughts @hongchaodeng @rynewang

File: src/ray/core_worker/test/fiber_state_test.cc [hunk 3]
```diff
@@ -95,6 +95,14 @@ TEST(FiberStateTest, RespectsConcurrencyLimit) {
   fiber_state.Join();
 }
 
+TEST(FiberStateTest, DoubleStopJoin) {
```
Comment (jjyao) [inline src/ray/core_worker/test/fiber_state_test.cc:98]: Besides the unit test, can we also add an e2e test using the repro script mentioned in the GH issue?

**Review by jjyao — APPROVED** [review]

File: src/ray/core_worker/test/fiber_state_test.cc (hunk 3 above)
Comment (hongchaodeng) [reply src/ray/core_worker/test/fiber_state_test.cc:98]: I don't think this is worth it.
The original repro script is sort of red herring. This test covers the root of the problem.

File: src/ray/core_worker/fiber.h (hunk 2 above)
Comment (hongchaodeng) [reply src/ray/core_worker/fiber.h:158]: Let's merge this and fix the issue first.
We are basically guaranteeing something that the library do not provide -- thread cancellation.
That's orthogonal to enhancing the capabilities.

File: src/ray/core_worker/test/fiber_state_test.cc (hunk 3 above)
Comment (jjyao) [reply src/ray/core_worker/test/fiber_state_test.cc:98]: I know this unit test covers the root cause of this issue, but it's still nice to make sure user's workload work well e2e even if, for example, we remove fiber completely in the future(and this unit test will be irrelevant)
